sdks/python/apache_beam/runners/portability/flink_runner_test.py

```diff
@@ -219,42 +223,72 @@ def test_flattened_side_input(self):
           with_transcoding=False)
 
     def test_metrics(self):
-      """Run a simple DoFn that increments a counter, and verify that its
-       expected value is written to a temporary file by the FileReporter"""
+      """Run a simple DoFn that increments a counter and verifies state
+      caching metrics. Verifies that its expected value is written to a
+      temporary file by the FileReporter"""
 
       counter_name = 'elem_counter'
+      state_spec = userstate.BagStateSpec('state', VarIntCoder())
 
       class DoFn(beam.DoFn):
         def __init__(self):
           self.counter = Metrics.counter(self.__class__, counter_name)
           logging.info('counter: %s' % self.counter.metric_name)
 
-        def process(self, v):
+        def process(self, kv, state=beam.DoFn.StateParam(state_spec)):
+          # Trigger materialization
+          list(state.read())
+          state.add(1)
           self.counter.inc()
 
       options = self.create_options()
       # Test only supports parallelism of 1
       options._all_options['parallelism'] = 1
-      n = 100
+      # Create multiple bundles to test cache metrics
+      options._all_options['max_bundle_size'] = 10
+      options._all_options['max_bundle_time_millis'] = 95130590130
+      experiments = options.view_as(DebugOptions).experiments or []
+      experiments.append('state_cache_size=123')
+      options.view_as(DebugOptions).experiments = experiments
       with Pipeline(self.get_runner(), options) as p:
         # pylint: disable=expression-not-assigned
         (p
-         | beam.Create(list(range(n)))
-         | beam.ParDo(DoFn()))
-
+         | "source" >> FlinkStreamingImpulseSource()
```

Comment:
Is it necessary to use the (deprecated) source here? Wouldn't `Create` also produce multiple bundles when  `max_bundle_size` is set?